pkg/mcs/port_set.go

```diff
@@ -0,0 +1,263 @@
+package mcs
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation"
+	mcsv1a1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+)
+
+const (
+	LIGHTHOUSE_PORT_KEY                   = "lighthouse.submariner.io/port"
+	LIGHTHOUSE_PORT_KEY_NAMED_FORMAT      = LIGHTHOUSE_PORT_KEY + ".%s"
+	LIGHTHOUSE_PORT_VALUE_FORMAT_PORT     = "%d"
+	LIGHTHOUSE_PORT_VALUE_FORMAT_PROTOCOL = LIGHTHOUSE_PORT_VALUE_FORMAT_PORT + "-%s"
+	LIGHTHOUSE_PORT_VALUE_FORMAT_FULL     = LIGHTHOUSE_PORT_VALUE_FORMAT_PROTOCOL + "-%s"
+)
+
+// PortSet encapsulates Servie ports used.
+type PortSet struct {
+	ports []mcsv1a1.ServicePort
+}
+
+// NewPortSet creates an empty PortSet.
+func NewPortSet() *PortSet {
+	return &PortSet{}
+}
+
+// NewPortSetFromServicePorts creates a PortSet populated from the ports defined
+// in v1.Service.Spec. Returns an error when the specification is invalid.
```

Comment:
Dont see that it can return any error

Reply:
Mostly correct for now, assuming we have a valid Service. This is mostly a convention that constructors should return an object and and error (e.g., what if the Service pointer is nil?)